ystia/yorc
# Feature/brbdcf 973 services support

# Pull Request description

## Description of the change

Support of Alien services in Deployment topologies :
the Deployment topology can reference a node template of an abstract node type (external service), or of an non-existent node type (managed service) which is actually a reference to a service provided by an application whose provisioning and lifecycle is not managed in this deployment.
The use of such an abstract or non-existent node type should is be accepted by the Orchestrator if the node template contains the directive substitutable.

Alien4Cloud is also today in this context of a service, managing capability attributes as node template attributes with the following naming convention : capabilities.capability_name.attribute_name.
For substitutable node templates, Yorc returns attribute values using this naming convention for attributes of capabilities exposed in a node template substitution mappings.
Similarly when asked to provide the value of a a substitutable node capability attribute, will return the value of the node template attribute of name capabilities.capability_name.attribute_name.

### What I did

Updated GetInstanceCapabilityAttribute() for a substitutable node to get capability attributes as node templates attributes with the format capabilities.capability_name.attribute_name,
updated as well this function for a specific case of attribute ip_address for a capability derived from an endpoint, assigned from the associated compute container.

Updated the definition store to store potential substitution_mappings stored in import files. This is used in the case of managed services as Alien4Cloud modifies the client deployment topology by defining a non-existent node type for the node template corresponding to the service, and adding an import of a topology whose template name is this non-existing type, and this topology imported contains substituion mapping that will allow us to know the real existing abstract node type of the node template. 
Storing also directives on node templates that the Yorc plugin defines to substitutable for node templates referencing services.

Updated GetInstanceAttribute() to take into account attributes following the naming scheme capabilities.capability_name.attribute_name : on substitutable nodes they refer to a node template attribute, on non-substitutable nodes (the source application providing a service) they refer to a capability attribute that Alien4Cloud fetches as attributes of a service and inject them as node templates attributes in client deployment templates referencing a managed Service.

Updated GetNodeType for substitutable nodes: if this type with no parent type, the real type is found in the import of a template whose template name is this type with no parent type

Adding the parsing of substitution mappings (Alien4Cloud doesn't support yet the definition of properties/attributes/interfaces mapping)

Misc.: Updated tosca topology template definition  (name, template, version defined in a section metadata). Updated public endpoints on k8s location to instantiate an attribute port to the value of the public port assigned by k8s for containers exposing their ports.


### How to verify it

This Pull Request has to be used with the corresponding Alien4Cloud Yorc plugin implementing feature BRBDCF-973-services-support

Setup :

In Alien4Cloud, upload a Yorc plugin from branch feature/BRBDCF-973-services-support
Install a Yorc orchestrator from this branch/
Define a Hosts Pool location of at least two hosts, each having labels public_adress and private_address

From Catalog > Manage archives > Git import > Git Location
Create a Git location :
Repository URL: https://github.com/ystia/yorc-a4c-plugin
Branch: feature/BRBDCF-973-services-support
Archive folder: tosca-samples/org/ystia/yorc/samples/consul

Create an application using the topology template ConsulServiceTemplate. At the last step of Deployment, select the checkbox allowing to create a Service in Alien4Cloud for this deployment.
Once the deployment is done, click on the output variable web_ui_url and check in the Nodes tab that there is just one node in the cluster.

Create an application using the topology ConsulClientTemplate. Check the topology to verify this topology is using an abstract node type for the Consul Server.
Check the Matching step and verify this abstract node type was matched against the service created in the previous step. Deploy the application.
Once the deployment is done, go back to the previous Application deployment, click on the output variable web_ui_url and check in the Nodes tab that there are now two nodes in the cluster.
